clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -488,7 +491,19 @@ private void close(KafkaChannel channel) {
         } catch (IOException e) {
             log.error("Exception closing connection to node {}:", channel.id(), e);
         }
-        this.stagedReceives.remove(channel);
+
+        // Keep track of closed channels with pending receives so that all received records
+        // may be processed. For example, when producer with acks=0 sends some records and
+        // closes its connections, a single poll() in the broker may receive records and
+        // handle close(). Closed channels are retained until the current poll() processing
+        // completes to enable all records received on the channel to be processed.
+        Deque<NetworkReceive> deque = this.stagedReceives.remove(channel);
+        if (deque != null) {
+            while (!deque.isEmpty()) {
+                addToCompletedReceives(channel, deque);
```

Comment:
Good point. It would be good to expand the following text in `poll` to explain why we only ever return one `completedReceive` per channel. It is something that has come up a few times.

```java
To overcome this issue we added "stagedReceives" map which contains per-channel deque. When we are
     * reading a channel we read as many responses as we can and store them into "stagedReceives" and pop one response during
     * the poll to add the completedReceives. If there are any active channels in the "stagedReceives" we set "timeout" to 0
     * and pop response and add to the completedReceives."
```